doc/config-generic-update-rollback/Json_Patch_Ordering_using_YANG_Models_Design.md

```diff
@@ -0,0 +1,703 @@
+- Each application of a JsonChange should move the configuration from a valid state to another valid state.
+- A valid state is a config that is valid according to YANG models.
+- Each JsonChange change should assume the applier of the change will have no notion of ordering. i.e. ordering of updates is only guaranteed between different JsonChange, within the same JsonChange order of update is not guaranteed.
+- It is assumed that the configuration lock was acquired before starting patch ordering.
+
+### 1.1.2 Configuration and Management Requirements
+N/A
+
+### 1.1.3 Scalability Requirements
+N/A
+
+### 1.1.4 Warm Boot Requirements
+N/A
+
+## 1.2 Design Overview
+
+From [SONiC Generic Configuration Update and Rollback - HLD](SONiC_Generic_Config_Update_and_Rollback_Design.md#3114-patch-orderer), this is the target interface:
+
+```python
+list<JsonChange> order-patch(JsonPatch jsonPatch)
+```
+
+|aspect      |item                 |description
+|------------|---------------------|-----------
+|inputs      |JsonPatch            | It represents the changes that needs to applied to the device running config, described in [JSON Patch (RFC6902)](https://tools.ietf.org/html/rfc6902).
+|outputs     |list&lt;JsonChange&gt;| The list will contain the steps to be followed to apply the input JsonPatch correctly. Each item in the list is assumed to be executed after the previous item, in the order given in the list.
+|errors      |malformedPatchError  | Will be raised if the input JsonPatch is not valid according to [JSON Patch (RFC6902)](https://tools.ietf.org/html/rfc6902).
+|            |other errors         | Check [SONiC Generic Configuration Update and Rollback - HLD](SONiC_Generic_Config_Update_and_Rollback_Design.md#3114-patch-orderer) for exact list of errors to expect.
+|side-effects|None                 |
+|assumptions |running-config locked| The implementor of this contract might interact with ConfigDB to get the running-config, it is assumed the running-config is locked for changes for the lifespan of the operation.
+
+The important constraint in the above interface is the JsonChange, where the application of a single JsonChange does not guarantee any ordering. If ordering is needed, multiple JsonChanges should be returned.
+
+This design document is exploring the option to use YANG models validations in order to verify each config state is valid.
+
+To represent this problem as a graph we will consider the following:
+- Nodes are configuration states which should be valid according to YANG models
+- Edges are the valid JsonChanges which should ignore any command ordering
+
+<img src="files/config-transitions-diagram.png" alt="config-transitions-diagram" width="1200"/>
+
+The problem now is just finding a path between the current config and the target config.  A solution to the above graph might be:
+
+<img src="files/valid-config-transitions-diagram.png" alt="valid-config-transitions-diagram" width="1200"/>
+
+Note:
```

Comment:
Think we can have a meeting about this because I feel this can be solved in a much simpler way.

From yang models, we can find the schema dependencies (i think we can also find it using pyang sonic yang tree generation). Using schema dependencies, we can create a graph based on the current config changes and then run topological sort on that.
But as said, This is not guaranteed that backend will process these config changes in order as I know. That is why we decided to depend on APP_DB and ASIC_DB updates in between config changes under dynamic Port breakout project.

Reply:
Booked a meeting for 3/15, main challenges to ordering are:
- Ordering divides the update into steps, each step need to move the config from a valid YANG model state to another YANG model state. Do we respect this constraint?
- Interdependency between network objects should be taken care of. For example: PORT/VLAN_MEMBER are related. Deleting a PORT/VLAN_MEMBER should start with deleting the VLAN_MEMBER then deleting the PORT. But adding PORT/VLAN_MEMBER should go in opposite order, adding the PORT adding the VLAN_MEMBER
- YANG models `mandatory` can force some fields to be required, in such case they cannot be deleted/added, only replaced.
- YANG models `min-elements` and `max-elements` can force the count of some items to be in certain range.
- YANG models `choice`/`unique` fields cannot be added then deleted as adding can create a conflict with existing fields, have to always be delete then added or replaced.
- All the above constraints can be combined

We separated apply-change mechanism to its own contract, so we can focus on ordering sedately from config application. (i.e. separation of concerns). Config application includes updating ConfigDB, restarting services, verifying services have absorbed change correctly, check Json_Change_Application_Design.md for details

Reply:
Offline discussion: The proposed approach takes into consideration using schema dependencies provided by YANG, but due to other requirements listed in the above reply, that won't be enough to allow the switch a solely topological sorting algorithm. It will be great if we can find a greedy solution to this problem, but so far there is not a clear one.